File: validSuduko.cpp
// Time Complexity :O(m * n) where m*n in the number elements in the matrix
// Space Complexity : O(m * n)  
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        int rows[9][9] ={0};
        int cols[9][9] = {0};
        int subBox[9][9] = {0};
        int k; int nums;
        for(int i=0;i< board.size();i++){
            for(int j=0;j< board[0].size();j++){
                if(board[i][j] != '.'){
                    nums = board[i][j] - '1';
                    k = i/3*3+j/3;
                    if(rows[i][nums] || cols[j][nums] || subBox[k][nums]) return false;
                    else {
                        rows[i][nums] = 1;
                        cols[j][nums] = 1;
                        subBox[k][nums] = 1;
                    }
                }
            }
        }
        return true;
    }
};

File: ValidSuduko.java
//Time complexity is O(M*N)
//Space complexity is 3*O(M*N)
class Solution {
    public boolean isValidSudoku(char[][] board) {
        int n =9;
        HashSet<Character>[] rows = new HashSet[n];
        HashSet<Character>[] cols = new HashSet[n];
        HashSet<Character>[] grids = new HashSet[n];
        for(int i=0;i<9;i++){
            for(int j=0;j<9;j++){
                Character val = board[i][j];
                
                if(board[i][j]!='.'){
                    if(rows[i]==null){
                        rows[i]=new HashSet<Character>();
                    }
                    if(rows[i].contains(val)){
                        return false;
                    }
                    else{
                        rows[i].add(val);
                    }
                    
                    
                    //Checking columns
                    if(cols[j]==null){
                        cols[j]=new HashSet<Character>();
                    }
                    if(cols[j].contains(val)){
                        return false;
                    }
                    else{
                        cols[j].add(val);
                    }
                    
                    //Checking grids
                    int gridnum = (i/3)*3+(j/3);
                    if(grids[gridnum]==null){
                        grids[gridnum]=new HashSet<Character>();
                    }
                    if(grids[gridnum].contains(val)){
                        return false;
                    }
                    else{
                        grids[gridnum].add(val);
                    }
                    
                }
            }
        }
        return true;
    }
}